### ISSUE TITLE
freeing memory with copy=false

### ISSUE THREAD

**Issue by beckerrh** [RUrlus/carma#19]
Hi Ralph,

I have a problem when using "col_to_arr" with copy=false, when simply passing an array back and forth  between python and C++.
There is py::capsule created which wants to free memory it didn't allocate, resulting into
" 
Python(9014,0x111c5edc0) malloc: *** error for object 0x1e4b6e000: pointer being freed was not allocated
"
If I uncomment the delete[] in the base-code, the code goes through without problem, but of coarse this is not what we want. I wonder if the 'delete' shouldn't only be executed if copy=true ?

Best regards,
Roland

**Comment by RUrlus** [comment]
Hi Roland,

The delete is needed to release the memory from python, regardless of copy. Without it there is a memory leak.

What is sounds like is that you try to do:
np.array (A) --> arma::Col --> np.array (B) without copying in or out?

If this is the case you are returning a reference to a buffer that is owned by array A which is a problem regardless of the arma object used in between. You should update array A after making changes rather than returning it as a new array.

An additional consideration is that stealing the memory of Row, Col or Cube is not easily done, see [Memory safety](https://carma.readthedocs.io/en/latest/memory_management.html#memory-safety) for details.

Unless you have strong reasons for using a Col over a (N x 1) matrix I would use the matrix for copy-less conversions

**Comment by beckerrh** [comment]
Hi Ralph,

Thanks a lot for your lightning fast answer.

> Le 27 mai 2020 à 14:26, Ralph Urlus <notifications@github.com> a écrit :
> Hi Roland,
> 
> The delete is needed to release the memory from python, regardless of copy. Without it there is a memory leak.
> 
My simple reasoning was that we do a delete on a pointer, for which we didn’t allocate memory ourselves in case of copy=true...
> What is sounds like is that you try to do:
> np.array (A) --> arma::Col --> np.array (B) without copying in or out?
> 
> 
Yes, exactly. Since NumPy and arma both can handle external memory, I was thinking, that would be possible (I understand that gets very complicated in case the memory is not aligned, but this is not the case in my application).
> If this is the case you are returning a reference to a buffer that is owned by array A which is a problem regardless of the arma object used in between. You should update array A after making changes rather than returning it as a new array.
> 
> An additional consideration is that stealing the memory of Row, Col or Cube is not easily done, see Memory safety <https://carma.readthedocs.io/en/latest/memory_management.html#memory-safety> for details.
> 
> Unless you have strong reasons for using a Col over a (N x 1) matrix I would use the matrix for copy-less conversions
> 
> 
Thanks a lot for these tips and for your documentation, which I will now study in detail. 

Best regards,
Roland

**Comment by RUrlus** [comment]
> My simple reasoning was that we do a delete on a pointer, for which we didn’t allocate memory ourselves in case of copy=true...

Yes we didn't allocate memory but we are stealing it from armadillo. We tell armadillo it's no longer owner of the memory and take the pointer. This way armadillo won't release the memory and we can control it.

> Yes, exactly. Since NumPy and arma both can handle external memory, I was thinking, that would be possible (I understand that gets very complicated in case the memory is not aligned, but this is not the case in my application). 

It's possible to do if you set owndata flag to false, but it's a bad idea. Now array A still owns the data and is unaware of the reference to its memory by B. So when A is garbage collected B will no longer point to valid memory.

This can all be avoided by simply updating the Numpy array you originally received, see [`update_array`](https://carma.readthedocs.io/en/latest/carma.html#update-array) for details

**Comment by beckerrh** [comment]
Hi Ralph,

Yes, I understand. But for some reason, it does not work the expected way for me.
I have made a small example showing my problem with a class just containing a arma::Mat<double> and set/get functions.
It works in the standard Cmake way (the py-file is copied), just
mkdir build; cd build; cmake ../carmatest; make; python3 carmatest.py 

Thanks a lot,
Roland

**Comment by RUrlus** [comment]
Where can I find this example?

**Comment by beckerrh** [comment]
Sorry, it was attached to my mail, but I didn't use GitHub, so it didn't appear. Here it is.
[carmatest.zip](https://github.com/RUrlus/carma/files/4695385/carmatest.zip)

**Comment by RUrlus** [comment]
No worries. To understand your use case:
You have a CPP class that you want to make accessible from python.

1. Set the data from python np array
2. Do something with the data in C++ using armadillo
3. Make the mutated data available in python as np array whilst keeping ownership of the data in CPP

Is this correct?

If so this use case is not directly supported at the moment but should be possible.

```c++
#include  <pybind11/pybind11.h>
#include  <pybind11/numpy.h>
#include  <carma/carma.h>

namespace py = pybind11;
/*=================================================*/
class carmatestpy
{
protected:
  arma::Mat<double> _test;
public:
  py::array_t<double> get_test()
{
  std::cerr << "get_solution() &_test = " << _test.memptr() << "\n";
  py::array_t<double> arr = carma::mat_to_arr<double>(_test);
  // inform numpy it does not own the buffer
  py::detail::array_proxy(arr.ptr())->flags &= ~py::detail::npy_api::NPY_ARRAY_OWNDATA_;
  // optionally make the array non-writable, other you will need to update arma matrix
  py::detail::array_proxy(arr.ptr())->flags &= ~py::detail::npy_api::NPY_ARRAY_WRITEABLE_;
  return arr;

}
void set_test(pybind11::array_t<double>& arr)
{
  _test =  carma::arr_to_mat<double>(arr);
  // inform numpy it no longer is owner of the buffer
  py::detail::array_proxy(arr.ptr())->flags &= ~py::detail::npy_api::NPY_ARRAY_OWNDATA_;
  // optionally prevent it being written to on the numpy side
  py::detail::array_proxy(arr.ptr())->flags &= ~py::detail::npy_api::NPY_ARRAY_WRITEABLE_;

  std::cerr << "set_solution() &_test = " << _test.memptr() << "\n";
}
};
```
This snippet is untested but should help you get there.
I'll see about getting the code to change the flags of the Numpy arrays in Carma.

Interestingly this is the opposite of how I use Carma, where python owns/controls the memory and we step into C++ for specific functions which then either return a new array or update it.

**Comment by RUrlus** [comment]
@beckerrh Did the above work for you?
I've added two functions for setting the flags, see 230f769c9b99337d50b8960cd93634e3eeef136d

**Comment by RUrlus** [comment]
Strange regarding regarding the flags.
I started building a class for your use-case that is intended to be inherited from, see 2f0661f5d6ad3c770cf07db04e6d0fe1230873fa, and the flags seem to be working correctly in this example.

**Comment by beckerrh** [comment]
> Strange regarding regarding the flags.
> 
No, I think flags are ok. The owndata -flag is false in both cases, either I set it or not (I guess because we have copy-false), whereas the value of the writable depends on whether I set it or not.
I guess the problem is that python still wants to delete the memory, even if the owndata-flag is wrong.

Also, I tried your class and run into the same problem.

**Comment by RUrlus** [comment]
Hi @beckerrh,

You were right, in the previous commit both the capsule and the matrix itself were trying to free the memory upon destruction when the memory was copied in.

The implementation is now done and merged into master, see bb9dac4f9f3e1879c3b01c5d1187d432bae80233
See the tests for an example of usage.

I'll close the issue, but feel free to reopen when you think it is not resolved.

Regards,

Ralph
